{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import nltk\n",
    "from tqdm import tqdm_notebook\n",
    "from collections import Iterable\n",
    "from itertools import chain\n",
    "import os\n",
    "import sqlite3\n",
    "import xmltodict\n",
    "from tqdm import tqdm_notebook\n",
    "from fuzzywuzzy import fuzz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect(\"human_protein.db\")\n",
    "prot_feat = pd.read_sql('select * from protein_features',con=conn)\n",
    "amy_df = pd.read_sql('select * from amyloid', con=conn)\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>accession</th>\n",
       "      <th>sequence</th>\n",
       "      <th>Nucleosome core</th>\n",
       "      <th>Calmodulin-binding</th>\n",
       "      <th>Cardiomyopathy</th>\n",
       "      <th>Electron transport</th>\n",
       "      <th>Glycerol metabolism</th>\n",
       "      <th>Heparin-binding</th>\n",
       "      <th>DNA integration</th>\n",
       "      <th>...</th>\n",
       "      <th>Iron-sulfur</th>\n",
       "      <th>Autism</th>\n",
       "      <th>Translocation</th>\n",
       "      <th>Signal transduction inhibitor</th>\n",
       "      <th>Ligand-gated ion channel</th>\n",
       "      <th>Phenylketonuria</th>\n",
       "      <th>Neuropathy</th>\n",
       "      <th>Hypogonadotropic hypogonadism</th>\n",
       "      <th>Aminotransferase</th>\n",
       "      <th>Multifunctional enzyme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RL37A_HUMAN</td>\n",
       "      <td>P61513,P12751,Q6FGF5</td>\n",
       "      <td>MAKRTKKVGIVGKYGTRYGASLRKMVKKIEISQHAKYTCSFCGKTK...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PYRG1_HUMAN</td>\n",
       "      <td>P17812,B4DR64,D3DPW1,Q5VW67,Q96GK6</td>\n",
       "      <td>MKYILVTGGVISGIGKGIIASSVGTILKSCGLHVTSIKIDPYINID...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RL6_HUMAN</td>\n",
       "      <td>Q02878,Q2M3Q3,Q8WW97</td>\n",
       "      <td>MAGEKVEKPDTKEKKPEAKKVDAGGKVKKGNLKAKKPKKGKPHCSR...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RAB10_HUMAN</td>\n",
       "      <td>P61026,D6W538,O88386,Q6IA52,Q9D7X6,Q9H0T3</td>\n",
       "      <td>MAKKTYDLLFKLLLIGDSGVGKTCVLFRFSDDAFNTTFISTIGIDF...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RAB30_HUMAN</td>\n",
       "      <td>Q15771,Q6FGK1,Q6MZH2,Q96CI8</td>\n",
       "      <td>MSMEDYDFLFKIVLIGNAGVGKTCLVRRFTQGLFPPGQGATIGVDF...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 828 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                                  accession  \\\n",
       "0  RL37A_HUMAN                       P61513,P12751,Q6FGF5   \n",
       "1  PYRG1_HUMAN         P17812,B4DR64,D3DPW1,Q5VW67,Q96GK6   \n",
       "2    RL6_HUMAN                       Q02878,Q2M3Q3,Q8WW97   \n",
       "3  RAB10_HUMAN  P61026,D6W538,O88386,Q6IA52,Q9D7X6,Q9H0T3   \n",
       "4  RAB30_HUMAN                Q15771,Q6FGK1,Q6MZH2,Q96CI8   \n",
       "\n",
       "                                            sequence   Nucleosome core  \\\n",
       "0  MAKRTKKVGIVGKYGTRYGASLRKMVKKIEISQHAKYTCSFCGKTK...                 0   \n",
       "1  MKYILVTGGVISGIGKGIIASSVGTILKSCGLHVTSIKIDPYINID...                 0   \n",
       "2  MAGEKVEKPDTKEKKPEAKKVDAGGKVKKGNLKAKKPKKGKPHCSR...                 0   \n",
       "3  MAKKTYDLLFKLLLIGDSGVGKTCVLFRFSDDAFNTTFISTIGIDF...                 0   \n",
       "4  MSMEDYDFLFKIVLIGNAGVGKTCLVRRFTQGLFPPGQGATIGVDF...                 0   \n",
       "\n",
       "    Calmodulin-binding   Cardiomyopathy   Electron transport  \\\n",
       "0                    0                0                    0   \n",
       "1                    0                0                    0   \n",
       "2                    0                0                    0   \n",
       "3                    0                0                    0   \n",
       "4                    0                0                    0   \n",
       "\n",
       "    Glycerol metabolism   Heparin-binding   DNA integration  \\\n",
       "0                     0                 0                 0   \n",
       "1                     0                 0                 0   \n",
       "2                     0                 0                 0   \n",
       "3                     0                 0                 0   \n",
       "4                     0                 0                 0   \n",
       "\n",
       "            ...              Iron-sulfur   Autism   Translocation  \\\n",
       "0           ...                        0        0               0   \n",
       "1           ...                        0        0               0   \n",
       "2           ...                        0        0               0   \n",
       "3           ...                        0        0               0   \n",
       "4           ...                        0        0               0   \n",
       "\n",
       "    Signal transduction inhibitor   Ligand-gated ion channel  \\\n",
       "0                               0                          0   \n",
       "1                               0                          0   \n",
       "2                               0                          0   \n",
       "3                               0                          0   \n",
       "4                               0                          0   \n",
       "\n",
       "    Phenylketonuria   Neuropathy   Hypogonadotropic hypogonadism  \\\n",
       "0                 0            0                               0   \n",
       "1                 0            0                               0   \n",
       "2                 0            0                               0   \n",
       "3                 0            0                               0   \n",
       "4                 0            0                               0   \n",
       "\n",
       "    Aminotransferase   Multifunctional enzyme  \n",
       "0                  0                        0  \n",
       "1                  0                        0  \n",
       "2                  0                        0  \n",
       "3                  0                        0  \n",
       "4                  0                        0  \n",
       "\n",
       "[5 rows x 828 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_feat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>species</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pmid</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AP00001</td>\n",
       "      <td>LPICPGGAARCQVTLRDLFDRAVVLSHYIHNLSSEMFSEFDKRYTH...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P01236</td>\n",
       "      <td>9006323</td>\n",
       "      <td>pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AP00002</td>\n",
       "      <td>CGNLSTCMLGTYTQDFNKFHTFPQTAIGVGAP</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P01258</td>\n",
       "      <td>12095997,17468747,23395606</td>\n",
       "      <td>pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AP00003</td>\n",
       "      <td>DEPPQSPWDRVKDLATVYVDVLKDSGRDYVSQFEGSALGKQLNLKL...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P02647</td>\n",
       "      <td>12421824,20182709,22609356</td>\n",
       "      <td>pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AP00004</td>\n",
       "      <td>KNTMEHVSSSEESIISQETYKQEKNMAINPSKENLCSTFCKEVVRN...</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>P02663</td>\n",
       "      <td>10611944</td>\n",
       "      <td>pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AP00005</td>\n",
       "      <td>RSFFSFLGEAFDGARDMWRAYSDMREANYIGSDKYFHARGNYDAAK...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P0DJI8</td>\n",
       "      <td>1731787,21569756</td>\n",
       "      <td>pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  amy_name                                           sequence       species  \\\n",
       "0  AP00001  LPICPGGAARCQVTLRDLFDRAVVLSHYIHNLSSEMFSEFDKRYTH...  Homo sapiens   \n",
       "1  AP00002                   CGNLSTCMLGTYTQDFNKFHTFPQTAIGVGAP  Homo sapiens   \n",
       "2  AP00003  DEPPQSPWDRVKDLATVYVDVLKDSGRDYVSQFEGSALGKQLNLKL...  Homo sapiens   \n",
       "3  AP00004  KNTMEHVSSSEESIISQETYKQEKNMAINPSKENLCSTFCKEVVRN...    Bos taurus   \n",
       "4  AP00005  RSFFSFLGEAFDGARDMWRAYSDMREANYIGSDKYFHARGNYDAAK...  Homo sapiens   \n",
       "\n",
       "  uniprot_id                        pmid    category  \n",
       "0     P01236                     9006323  pathogenic  \n",
       "1     P01258  12095997,17468747,23395606  pathogenic  \n",
       "2     P02647  12421824,20182709,22609356  pathogenic  \n",
       "3     P02663                    10611944  pathogenic  \n",
       "4     P0DJI8            1731787,21569756  pathogenic  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amy_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Checking amyloid status against uniprot proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(71, 72)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def idChecker(accessions, uni_id):\n",
    "    if uni_id in accessions:\n",
    "        return True\n",
    "    return False\n",
    "# Split the accession strings to a list\n",
    "split_accs = prot_feat['accession'].str.split(',').values.tolist()\n",
    "# Match a list out of the set to check against in the checker func\n",
    "acc_list = list(set([item for sublist in split_accs for item in sublist]))\n",
    "# Add column to hold info on match for each amyloid\n",
    "amy_df.loc[:,'uni_match'] = amy_df.uniprot_id.apply(lambda x: idChecker(acc_list, x)).values\n",
    "# Amyloids with no match\n",
    "no_uni_amys = amy_df[~amy_df.uni_match]\n",
    "# Amyloids with a match\n",
    "match_amys = amy_df[amy_df.uni_match]\n",
    "# Num amyloids that have a match, num amyloids with no match\n",
    "amy_df.shape[0]-no_uni_amys.shape[0], amy_df.shape[0]-match_amys.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So there are 72 proteins in the amyloid table that have no representation in the protein list. Those proteins are going to be missing their structural information and other annotations from uniprot.\n",
    "\n",
    "I want to avoid missing out on all the really nice annotation information held in each of the annotated proteins. So I'm going to request the uniprot XML information for each of the amyloid proteins that are missing in the reviewed section of uniprot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_ids = no_uni_amys.uniprot_id.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'P02663 P02754 P63159 P32081 P0C805 P68082 Q52546 P05453 P23202 Q03689 P23727 P28307 P0ABK7 Q9AD92 Q9Z4N4 Q04571 P25367 P14922 P09547 P02846 P04002 Q9PWC8 P07884 P54785 Q08972 Q02629 P40070 P52912 P38996 Q967R6 A8Z0V1 Q53643 P55090 P01144 P01145 P10636-8 A0MVU0 P11657 Q9VSR3 P0A734 Q59L12 P82042 P32194 P59637 P25822 P03275 P0ADA7 P61825 Q4ZHU1 Q1EN15 P01012 P00698 J7GMN2 P03036 P86706 P15703 P54507 D2YW43 P52750 P32588 P38741 C4IN70 C4IN69 A0B829 P0C0U1 A0A1A0G6M5 Q99109 Q934F8 P53617 P08615 P08611 '"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_str = \" \".join(missing_ids)\n",
    "copy_str"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check matched amyloid sequence against annotated sequence.\n",
    "\n",
    "If my amyloid protein sequences are the same as the protein with the same uniprot ID in my protein table - then I can avoid having to add the amyloid sequence into the db.\n",
    "\n",
    "If the amyloid sequence is different, I will make a 'copy' row that contains the variant amyloid sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First lets get the table of protein name and accession melted on accesion.\n",
    "prot_feat.loc[:,'accession_split'] = prot_feat['accession'].str.split(',')\n",
    "prot_feat_extended = prot_feat[['protein','sequence','accession_split']] \\\n",
    "    .set_index(['protein','sequence'])['accession_split'].apply(pd.Series).stack()\n",
    "prot_feat_extended = prot_feat_extended.reset_index()\n",
    "prot_feat_extended.columns=['protein','sequence','id_num','uniprot_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence</th>\n",
       "      <th>id_num</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RL37A_HUMAN</td>\n",
       "      <td>MAKRTKKVGIVGKYGTRYGASLRKMVKKIEISQHAKYTCSFCGKTK...</td>\n",
       "      <td>0</td>\n",
       "      <td>P61513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RL37A_HUMAN</td>\n",
       "      <td>MAKRTKKVGIVGKYGTRYGASLRKMVKKIEISQHAKYTCSFCGKTK...</td>\n",
       "      <td>1</td>\n",
       "      <td>P12751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RL37A_HUMAN</td>\n",
       "      <td>MAKRTKKVGIVGKYGTRYGASLRKMVKKIEISQHAKYTCSFCGKTK...</td>\n",
       "      <td>2</td>\n",
       "      <td>Q6FGF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PYRG1_HUMAN</td>\n",
       "      <td>MKYILVTGGVISGIGKGIIASSVGTILKSCGLHVTSIKIDPYINID...</td>\n",
       "      <td>0</td>\n",
       "      <td>P17812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PYRG1_HUMAN</td>\n",
       "      <td>MKYILVTGGVISGIGKGIIASSVGTILKSCGLHVTSIKIDPYINID...</td>\n",
       "      <td>1</td>\n",
       "      <td>B4DR64</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                                           sequence  id_num  \\\n",
       "0  RL37A_HUMAN  MAKRTKKVGIVGKYGTRYGASLRKMVKKIEISQHAKYTCSFCGKTK...       0   \n",
       "1  RL37A_HUMAN  MAKRTKKVGIVGKYGTRYGASLRKMVKKIEISQHAKYTCSFCGKTK...       1   \n",
       "2  RL37A_HUMAN  MAKRTKKVGIVGKYGTRYGASLRKMVKKIEISQHAKYTCSFCGKTK...       2   \n",
       "3  PYRG1_HUMAN  MKYILVTGGVISGIGKGIIASSVGTILKSCGLHVTSIKIDPYINID...       0   \n",
       "4  PYRG1_HUMAN  MKYILVTGGVISGIGKGIIASSVGTILKSCGLHVTSIKIDPYINID...       1   \n",
       "\n",
       "  uniprot_id  \n",
       "0     P61513  \n",
       "1     P12751  \n",
       "2     Q6FGF5  \n",
       "3     P17812  \n",
       "4     B4DR64  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_feat_extended.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge on uniprot ID\n",
    "prot_amy_merge = prot_feat_extended[['protein','sequence','uniprot_id']] \\\n",
    "    .merge(amy_df[['amy_name','sequence','uniprot_id','species']],\n",
    "          on='uniprot_id',how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(15, 56)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_amy_merge.head()\n",
    "# Check for sequence equivalence\n",
    "seq_equals = prot_amy_merge['sequence_x'].str.strip()==prot_amy_merge['sequence_y'].str.strip()\n",
    "print(seq_equals.shape[0])\n",
    "# Num sequences equivalent subtracted by sequences with no equivalence\n",
    "prot_amy_merge[seq_equals].shape[0],prot_amy_merge[~seq_equals].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSPC_HUMAN</td>\n",
       "      <td>MDVGSKEVLMESPPDYSAAPRGRFGIPCCPVHLKRLLIVVVVVVLI...</td>\n",
       "      <td>P11686</td>\n",
       "      <td>AP00130</td>\n",
       "      <td>FGIPCCPVHLKRLLIVVVVVVLIVVVIVGALLMGL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SODC_HUMAN</td>\n",
       "      <td>MATKAVCVLKGDGPVQGIINFEQKESNGPVKVWGSIKGLTEGLHGF...</td>\n",
       "      <td>P00441</td>\n",
       "      <td>AP00084</td>\n",
       "      <td>ATKAVCVLKGDGPVQGIINFEQKESNGPVKVWGSIKGLTEGLHGFH...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAA1_HUMAN</td>\n",
       "      <td>MKLLTGLVFCSLVLGVSSRSFFSFLGEAFDGARDMWRAYSDMREAN...</td>\n",
       "      <td>P0DJI8</td>\n",
       "      <td>AP00005</td>\n",
       "      <td>RSFFSFLGEAFDGARDMWRAYSDMREANYIGSDKYFHARGNYDAAK...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SEMG1_HUMAN</td>\n",
       "      <td>MKPNIIFVLSLLLILEKQAAVMGQKGGSKGRLPSEFSQFPHGQKGQ...</td>\n",
       "      <td>P04279</td>\n",
       "      <td>AP00009</td>\n",
       "      <td>QKGGSKGRLPSEFSQFPHGQKGQHYSGQKGKQQTESKGSFSIQYTY...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SOMA_HUMAN</td>\n",
       "      <td>MATGSRTSLLLAFGLLCLPWLQEGSAFPTIPLSRLFDNAMLRAHRL...</td>\n",
       "      <td>P01241</td>\n",
       "      <td>AP00135</td>\n",
       "      <td>FPTIPLSRLFDNAMLRAHRLHQLAFDTYQEFEEAYIPKEQKYSFLQ...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                                         sequence_x uniprot_id  \\\n",
       "0   PSPC_HUMAN  MDVGSKEVLMESPPDYSAAPRGRFGIPCCPVHLKRLLIVVVVVVLI...     P11686   \n",
       "1   SODC_HUMAN  MATKAVCVLKGDGPVQGIINFEQKESNGPVKVWGSIKGLTEGLHGF...     P00441   \n",
       "2   SAA1_HUMAN  MKLLTGLVFCSLVLGVSSRSFFSFLGEAFDGARDMWRAYSDMREAN...     P0DJI8   \n",
       "4  SEMG1_HUMAN  MKPNIIFVLSLLLILEKQAAVMGQKGGSKGRLPSEFSQFPHGQKGQ...     P04279   \n",
       "5   SOMA_HUMAN  MATGSRTSLLLAFGLLCLPWLQEGSAFPTIPLSRLFDNAMLRAHRL...     P01241   \n",
       "\n",
       "  amy_name                                         sequence_y       species  \n",
       "0  AP00130                FGIPCCPVHLKRLLIVVVVVVLIVVVIVGALLMGL  Homo sapiens  \n",
       "1  AP00084  ATKAVCVLKGDGPVQGIINFEQKESNGPVKVWGSIKGLTEGLHGFH...  Homo sapiens  \n",
       "2  AP00005  RSFFSFLGEAFDGARDMWRAYSDMREANYIGSDKYFHARGNYDAAK...  Homo sapiens  \n",
       "4  AP00009  QKGGSKGRLPSEFSQFPHGQKGQHYSGQKGKQQTESKGSFSIQYTY...  Homo sapiens  \n",
       "5  AP00135  FPTIPLSRLFDNAMLRAHRLHQLAFDTYQEFEEAYIPKEQKYSFLQ...  Homo sapiens  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_amy_merge[~seq_equals].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So I've got 41 more uniprot-amyloid matches with no sequence similarity. Lets see if the sequence is simply truncated and exists within the uniprot sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(56, 6)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(48, 6)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_merge_no_match = prot_amy_merge[~seq_equals]\n",
    "print(prot_merge_no_match.shape)\n",
    "def stringIncluded(str1, str2):\n",
    "    if str1 in str2 or str1 == str2:\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "string_inside_index = prot_merge_no_match.apply(lambda row: stringIncluded(row['sequence_y'],row['sequence_x']) ,axis=1)\n",
    "string_inside = prot_merge_no_match[string_inside_index]\n",
    "string_not_inside = prot_merge_no_match[~string_inside_index]\n",
    "string_inside.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "100\n",
      "False\n",
      "98\n",
      "False\n",
      "88\n",
      "False\n",
      "19\n",
      "False\n",
      "17\n",
      "False\n",
      "98\n",
      "False\n",
      "97\n",
      "False\n",
      "78\n"
     ]
    }
   ],
   "source": [
    "for x, y in string_not_inside[['sequence_x','sequence_y']].values:\n",
    "    print(x==y)\n",
    "    print(fuzz.ratio(x,y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TADBP_HUMAN</td>\n",
       "      <td>MSEYIRVTEDENDEPIEIPSEDDGTVLLSTVTAQFPGACGLRYRNP...</td>\n",
       "      <td>Q13148</td>\n",
       "      <td>AP00077</td>\n",
       "      <td>MSEYIRVTEDENDEPIEIPSEDDGTVLLSTVTAQFPGACGLRYRNP...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BGH3_HUMAN</td>\n",
       "      <td>MALFVRLLALALALALGPAATLAGPAKSPYQLVLQHSRLRGRQHGP...</td>\n",
       "      <td>Q15582</td>\n",
       "      <td>AP00095</td>\n",
       "      <td>GPAKSPYQLVLQHSRLRGRQHGPNVCAVQKVIGTNRKYFTNCKQWY...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>APOC3_HUMAN</td>\n",
       "      <td>MQPRVLLVVALLALLASARASEAEDASLLSFMQGYMKHATKTAKDA...</td>\n",
       "      <td>P02656</td>\n",
       "      <td>AP00142</td>\n",
       "      <td>SEAEDASLLSFMQGYMKHATKTAKVALSSVQESQVAQQARGWVTDG...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ITM2B_HUMAN</td>\n",
       "      <td>MVKVTFNSALAQKEAKKDEPKSGEEALIIPPDAVAVDCKDPDDVVP...</td>\n",
       "      <td>Q9Y287</td>\n",
       "      <td>AP00107</td>\n",
       "      <td>EASNCFAIRHFENKFAVETLICSRTVKKNIIEEN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ITM2B_HUMAN</td>\n",
       "      <td>MVKVTFNSALAQKEAKKDEPKSGEEALIIPPDAVAVDCKDPDDVVP...</td>\n",
       "      <td>Q9Y287</td>\n",
       "      <td>AP00108</td>\n",
       "      <td>EASNCFAIRHFENKFAVETLICFNLFLNSQEKHY</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        protein                                         sequence_x uniprot_id  \\\n",
       "9   TADBP_HUMAN  MSEYIRVTEDENDEPIEIPSEDDGTVLLSTVTAQFPGACGLRYRNP...     Q13148   \n",
       "12   BGH3_HUMAN  MALFVRLLALALALALGPAATLAGPAKSPYQLVLQHSRLRGRQHGP...     Q15582   \n",
       "14  APOC3_HUMAN  MQPRVLLVVALLALLASARASEAEDASLLSFMQGYMKHATKTAKDA...     P02656   \n",
       "17  ITM2B_HUMAN  MVKVTFNSALAQKEAKKDEPKSGEEALIIPPDAVAVDCKDPDDVVP...     Q9Y287   \n",
       "18  ITM2B_HUMAN  MVKVTFNSALAQKEAKKDEPKSGEEALIIPPDAVAVDCKDPDDVVP...     Q9Y287   \n",
       "\n",
       "   amy_name                                         sequence_y       species  \n",
       "9   AP00077  MSEYIRVTEDENDEPIEIPSEDDGTVLLSTVTAQFPGACGLRYRNP...  Homo sapiens  \n",
       "12  AP00095  GPAKSPYQLVLQHSRLRGRQHGPNVCAVQKVIGTNRKYFTNCKQWY...  Homo sapiens  \n",
       "14  AP00142  SEAEDASLLSFMQGYMKHATKTAKVALSSVQESQVAQQARGWVTDG...  Homo sapiens  \n",
       "17  AP00107                 EASNCFAIRHFENKFAVETLICSRTVKKNIIEEN  Homo sapiens  \n",
       "18  AP00108                 EASNCFAIRHFENKFAVETLICFNLFLNSQEKHY  Homo sapiens  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "string_not_inside.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I'm going to just avoid using these in any train/test set. I can use them after the fact to check how well the model did in predicting the uniprot annotated version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect('human_protein.db')\n",
    "string_not_inside[['protein','uniprot_id','amy_name']].to_sql(\"amy_ignore\",con=conn,index=False,if_exists='replace')\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Correct amyloid_prion mapping\n",
    "\n",
    "For the amyloids that exist for humans and are simply noted as truncated, alter the coordinates of their prions so that they map to the uniprot sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(174, 3)\n",
      "(174, 4)\n"
     ]
    }
   ],
   "source": [
    "# Alter amyloid_prion such that;\n",
    "# All entries have uniprot_id as their primary key\n",
    "conn = sqlite3.connect('human_protein.db')\n",
    "amy_uni_ids = pd.read_sql(\"select amy_name, uniprot_id from amyloid\",con=conn)\n",
    "amy_prion = pd.read_sql('select * from amyloid_prion',con=conn)\n",
    "print(amy_prion.shape)\n",
    "amy_joined = amy_uni_ids.merge(amy_prion, on='amy_name', how='right')\n",
    "print(amy_joined.shape)\n",
    "amy_joined.drop('amy_name',axis=1,inplace=True)\n",
    "# There is one entry of a synthetic protein. Drop it.\n",
    "amy_joined = amy_joined[amy_joined['uniprot_id'].str.len()!=0]\n",
    "amy_joined.to_sql(\"amyloid_prion_id\",con=conn,index=False,if_exists='replace')\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(  uniprot_id begin end\n",
       " 0     P01236     7  34\n",
       " 1     P01236    43  57\n",
       " 2     P01258     6  11\n",
       " 3     P01258    15  20\n",
       " 4     P02647    46  59, (173, 3))"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ensure human proteins that are truncated, get mapped to uniprot sequence.\n",
    "conn = sqlite3.connect('human_protein.db')\n",
    "amy_prion_df = pd.read_sql(\"select * from amyloid_prion_id\",con=conn)\n",
    "conn.close()\n",
    "amy_prion_df.head(), amy_prion_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSPC_HUMAN</td>\n",
       "      <td>MDVGSKEVLMESPPDYSAAPRGRFGIPCCPVHLKRLLIVVVVVVLI...</td>\n",
       "      <td>P11686</td>\n",
       "      <td>AP00130</td>\n",
       "      <td>FGIPCCPVHLKRLLIVVVVVVLIVVVIVGALLMGL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SODC_HUMAN</td>\n",
       "      <td>MATKAVCVLKGDGPVQGIINFEQKESNGPVKVWGSIKGLTEGLHGF...</td>\n",
       "      <td>P00441</td>\n",
       "      <td>AP00084</td>\n",
       "      <td>ATKAVCVLKGDGPVQGIINFEQKESNGPVKVWGSIKGLTEGLHGFH...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAA1_HUMAN</td>\n",
       "      <td>MKLLTGLVFCSLVLGVSSRSFFSFLGEAFDGARDMWRAYSDMREAN...</td>\n",
       "      <td>P0DJI8</td>\n",
       "      <td>AP00005</td>\n",
       "      <td>RSFFSFLGEAFDGARDMWRAYSDMREANYIGSDKYFHARGNYDAAK...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SEMG1_HUMAN</td>\n",
       "      <td>MKPNIIFVLSLLLILEKQAAVMGQKGGSKGRLPSEFSQFPHGQKGQ...</td>\n",
       "      <td>P04279</td>\n",
       "      <td>AP00009</td>\n",
       "      <td>QKGGSKGRLPSEFSQFPHGQKGQHYSGQKGKQQTESKGSFSIQYTY...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SOMA_HUMAN</td>\n",
       "      <td>MATGSRTSLLLAFGLLCLPWLQEGSAFPTIPLSRLFDNAMLRAHRL...</td>\n",
       "      <td>P01241</td>\n",
       "      <td>AP00135</td>\n",
       "      <td>FPTIPLSRLFDNAMLRAHRLHQLAFDTYQEFEEAYIPKEQKYSFLQ...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                                         sequence_x uniprot_id  \\\n",
       "0   PSPC_HUMAN  MDVGSKEVLMESPPDYSAAPRGRFGIPCCPVHLKRLLIVVVVVVLI...     P11686   \n",
       "1   SODC_HUMAN  MATKAVCVLKGDGPVQGIINFEQKESNGPVKVWGSIKGLTEGLHGF...     P00441   \n",
       "2   SAA1_HUMAN  MKLLTGLVFCSLVLGVSSRSFFSFLGEAFDGARDMWRAYSDMREAN...     P0DJI8   \n",
       "4  SEMG1_HUMAN  MKPNIIFVLSLLLILEKQAAVMGQKGGSKGRLPSEFSQFPHGQKGQ...     P04279   \n",
       "5   SOMA_HUMAN  MATGSRTSLLLAFGLLCLPWLQEGSAFPTIPLSRLFDNAMLRAHRL...     P01241   \n",
       "\n",
       "  amy_name                                         sequence_y       species  \\\n",
       "0  AP00130                FGIPCCPVHLKRLLIVVVVVVLIVVVIVGALLMGL  Homo sapiens   \n",
       "1  AP00084  ATKAVCVLKGDGPVQGIINFEQKESNGPVKVWGSIKGLTEGLHGFH...  Homo sapiens   \n",
       "2  AP00005  RSFFSFLGEAFDGARDMWRAYSDMREANYIGSDKYFHARGNYDAAK...  Homo sapiens   \n",
       "4  AP00009  QKGGSKGRLPSEFSQFPHGQKGQHYSGQKGKQQTESKGSFSIQYTY...  Homo sapiens   \n",
       "5  AP00135  FPTIPLSRLFDNAMLRAHRLHQLAFDTYQEFEEAYIPKEQKYSFLQ...  Homo sapiens   \n",
       "\n",
       "   seq_diff  \n",
       "0        23  \n",
       "1         1  \n",
       "2        18  \n",
       "4        23  \n",
       "5        26  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(None, (48, 7))"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display(string_inside.head()), string_inside.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [protein, sequence_x, uniprot_id, amy_name, sequence_y, species, seq_diff]\n",
       "Index: []"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "string_inside[string_inside.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Jake\\Anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py:543: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[item] = s\n"
     ]
    }
   ],
   "source": [
    "def prionAnnotMapper(seq1, seq2):\n",
    "    \"\"\"Return difference between the start of seq1 and where seq2 resides\n",
    "    within seq1. This will be used to alter the begin and end columns so \n",
    "    that each amyloid annotated region can be mapped to its uniprot\n",
    "    counterpart.\"\"\"\n",
    "    return len(seq1.split(seq2)[0])\n",
    "\n",
    "# Apply prion annot mapper to each row in amy_prion_df that has an entry in\n",
    "# the string_inside dataframe.\n",
    "string_inside.loc[:,'seq_diff'] = string_inside.apply(lambda row: prionAnnotMapper(row['sequence_x'],row['sequence_y']),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GLUC_HUMAN</td>\n",
       "      <td>MKSIYFVAGLFVMLVQGSWQRSLQDTEEKSRSFSASQADPLSDPDQ...</td>\n",
       "      <td>P01275</td>\n",
       "      <td>AP00041</td>\n",
       "      <td>HSQGTFTSDYSKYLDSRRAQDFVQWLMNT</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GLUC_HUMAN</td>\n",
       "      <td>MKSIYFVAGLFVMLVQGSWQRSLQDTEEKSRSFSASQADPLSDPDQ...</td>\n",
       "      <td>P01275</td>\n",
       "      <td>AP00065</td>\n",
       "      <td>HDEFERHAEGTFTSDVSSYLEGQAAKEFIAWLVKGRG</td>\n",
       "      <td>Home sapiens</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GLUC_HUMAN</td>\n",
       "      <td>MKSIYFVAGLFVMLVQGSWQRSLQDTEEKSRSFSASQADPLSDPDQ...</td>\n",
       "      <td>P01275</td>\n",
       "      <td>AP00066</td>\n",
       "      <td>HADGSFSDEMNTILDNLAARDFINWLIQTKITD</td>\n",
       "      <td>Home sapiens</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                                         sequence_x uniprot_id  \\\n",
       "19  GLUC_HUMAN  MKSIYFVAGLFVMLVQGSWQRSLQDTEEKSRSFSASQADPLSDPDQ...     P01275   \n",
       "20  GLUC_HUMAN  MKSIYFVAGLFVMLVQGSWQRSLQDTEEKSRSFSASQADPLSDPDQ...     P01275   \n",
       "21  GLUC_HUMAN  MKSIYFVAGLFVMLVQGSWQRSLQDTEEKSRSFSASQADPLSDPDQ...     P01275   \n",
       "\n",
       "   amy_name                             sequence_y       species  seq_diff  \n",
       "19  AP00041          HSQGTFTSDYSKYLDSRRAQDFVQWLMNT  Homo sapiens        52  \n",
       "20  AP00065  HDEFERHAEGTFTSDVSSYLEGQAAKEFIAWLVKGRG  Home sapiens        91  \n",
       "21  AP00066      HADGSFSDEMNTILDNLAARDFINWLIQTKITD  Home sapiens       145  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>INS_HUMAN</td>\n",
       "      <td>MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGER...</td>\n",
       "      <td>P01308</td>\n",
       "      <td>AP00023</td>\n",
       "      <td>FVNQHLCGSHLVEALYLVCGERGFFYTPKT</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>INS_HUMAN</td>\n",
       "      <td>MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGER...</td>\n",
       "      <td>P01308</td>\n",
       "      <td>AP00024</td>\n",
       "      <td>GIVEQCCTSICSLYQLENYCN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein                                         sequence_x uniprot_id  \\\n",
       "22  INS_HUMAN  MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGER...     P01308   \n",
       "23  INS_HUMAN  MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGER...     P01308   \n",
       "\n",
       "   amy_name                      sequence_y       species  seq_diff  \n",
       "22  AP00023  FVNQHLCGSHLVEALYLVCGERGFFYTPKT  Homo sapiens        24  \n",
       "23  AP00024           GIVEQCCTSICSLYQLENYCN  Homo sapiens        89  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>PPAP_HUMAN</td>\n",
       "      <td>MRAAPLLLARAASLSLGFLFLLFFWLDRSVLAKELKFVTLVFRHGD...</td>\n",
       "      <td>P15309</td>\n",
       "      <td>AP00091</td>\n",
       "      <td>GIHKQKEKSRLQGGVLVNEILNHMKRATQIPSYKKLIMY</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>PPAP_HUMAN</td>\n",
       "      <td>MRAAPLLLARAASLSLGFLFLLFFWLDRSVLAKELKFVTLVFRHGD...</td>\n",
       "      <td>P15309</td>\n",
       "      <td>AP00092</td>\n",
       "      <td>IRKRYRKFLNESYKHEQVYIRSTDVDRTLMSAMTNL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                                         sequence_x uniprot_id  \\\n",
       "53  PPAP_HUMAN  MRAAPLLLARAASLSLGFLFLLFFWLDRSVLAKELKFVTLVFRHGD...     P15309   \n",
       "54  PPAP_HUMAN  MRAAPLLLARAASLSLGFLFLLFFWLDRSVLAKELKFVTLVFRHGD...     P15309   \n",
       "\n",
       "   amy_name                               sequence_y       species  seq_diff  \n",
       "53  AP00091  GIHKQKEKSRLQGGVLVNEILNHMKRATQIPSYKKLIMY  Homo sapiens       247  \n",
       "54  AP00092     IRKRYRKFLNESYKHEQVYIRSTDVDRTLMSAMTNL  Homo sapiens        84  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>ANF_HUMAN</td>\n",
       "      <td>MSSFSTTTVSFLLLLAFQLLGQTRANPMYNAVSNADLMDFKNLLDH...</td>\n",
       "      <td>P01160</td>\n",
       "      <td>AP00081</td>\n",
       "      <td>NPMYNAVSNADLMDFKNLLDHLEEKMPLEDEVVPPQVLSEPNEEAG...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>ANF_HUMAN</td>\n",
       "      <td>MSSFSTTTVSFLLLLAFQLLGQTRANPMYNAVSNADLMDFKNLLDH...</td>\n",
       "      <td>P01160</td>\n",
       "      <td>AP00082</td>\n",
       "      <td>SLRRSSCFGGRMDRIGAQSGLGCNSFRY</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein                                         sequence_x uniprot_id  \\\n",
       "62  ANF_HUMAN  MSSFSTTTVSFLLLLAFQLLGQTRANPMYNAVSNADLMDFKNLLDH...     P01160   \n",
       "63  ANF_HUMAN  MSSFSTTTVSFLLLLAFQLLGQTRANPMYNAVSNADLMDFKNLLDH...     P01160   \n",
       "\n",
       "   amy_name                                         sequence_y       species  \\\n",
       "62  AP00081  NPMYNAVSNADLMDFKNLLDHLEEKMPLEDEVVPPQVLSEPNEEAG...  Homo sapiens   \n",
       "63  AP00082                       SLRRSSCFGGRMDRIGAQSGLGCNSFRY  Homo sapiens   \n",
       "\n",
       "    seq_diff  \n",
       "62        25  \n",
       "63       123  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>ANFB_HUMAN</td>\n",
       "      <td>MDPQTAPSRALLLLLFLHLAFLGGRSHPLGSPGSASDLETSGLQEQ...</td>\n",
       "      <td>P16860</td>\n",
       "      <td>AP00132</td>\n",
       "      <td>HPLGSPGSASDLETSGLQEQRNHLQGKLSELQVEQTSLEPLQESPR...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>ANFB_HUMAN</td>\n",
       "      <td>MDPQTAPSRALLLLLFLHLAFLGGRSHPLGSPGSASDLETSGLQEQ...</td>\n",
       "      <td>P16860</td>\n",
       "      <td>AP00019</td>\n",
       "      <td>SPKMVQGSGCFGRKMDRISSSSGLGCKVLRRH</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                                         sequence_x uniprot_id  \\\n",
       "66  ANFB_HUMAN  MDPQTAPSRALLLLLFLHLAFLGGRSHPLGSPGSASDLETSGLQEQ...     P16860   \n",
       "67  ANFB_HUMAN  MDPQTAPSRALLLLLFLHLAFLGGRSHPLGSPGSASDLETSGLQEQ...     P16860   \n",
       "\n",
       "   amy_name                                         sequence_y       species  \\\n",
       "66  AP00132  HPLGSPGSASDLETSGLQEQRNHLQGKLSELQVEQTSLEPLQESPR...  Homo sapiens   \n",
       "67  AP00019                   SPKMVQGSGCFGRKMDRISSSSGLGCKVLRRH  Homo sapiens   \n",
       "\n",
       "    seq_diff  \n",
       "66        26  \n",
       "67       102  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>sequence_x</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>amy_name</th>\n",
       "      <th>sequence_y</th>\n",
       "      <th>species</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>A4_HUMAN</td>\n",
       "      <td>MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...</td>\n",
       "      <td>P05067</td>\n",
       "      <td>AP00010</td>\n",
       "      <td>DAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVIA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>A4_HUMAN</td>\n",
       "      <td>MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...</td>\n",
       "      <td>P05067</td>\n",
       "      <td>AP00011</td>\n",
       "      <td>DAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVV</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>671</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     protein                                         sequence_x uniprot_id  \\\n",
       "68  A4_HUMAN  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...     P05067   \n",
       "69  A4_HUMAN  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...     P05067   \n",
       "\n",
       "   amy_name                                  sequence_y       species  \\\n",
       "68  AP00010  DAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVIA  Homo sapiens   \n",
       "69  AP00011    DAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVV  Homo sapiens   \n",
       "\n",
       "    seq_diff  \n",
       "68       671  \n",
       "69       671  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "checked_id = set()\n",
    "for uni_id in string_inside['uniprot_id']:\n",
    "    if uni_id not in checked_id:\n",
    "        checked_id.add(uni_id)\n",
    "        id_slice = string_inside[string_inside['uniprot_id']==uni_id]\n",
    "        if id_slice.shape[0]>1:\n",
    "            display(id_slice)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(173, 3) (187, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P01236</td>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P01236</td>\n",
       "      <td>43</td>\n",
       "      <td>57</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P01258</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P01258</td>\n",
       "      <td>15</td>\n",
       "      <td>20</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P02647</td>\n",
       "      <td>46</td>\n",
       "      <td>59</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id begin end  seq_diff\n",
       "0     P01236     7  34      28.0\n",
       "1     P01236    43  57      28.0\n",
       "2     P01258     6  11      84.0\n",
       "3     P01258    15  20      84.0\n",
       "4     P02647    46  59      24.0"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amy_prion_df_alt = amy_prion_df.merge(string_inside[['uniprot_id','seq_diff']],on='uniprot_id',how='left').drop_duplicates()\n",
    "amy_prion_df_alt.fillna(0,inplace=True)\n",
    "print(amy_prion_df.shape,amy_prion_df_alt.shape)\n",
    "amy_prion_df_alt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "amy_prion_df_alt.loc[:,'begin_diff'] = amy_prion_df_alt['begin'].astype(int) + amy_prion_df_alt['seq_diff'].astype(int)\n",
    "amy_prion_df_alt.loc[:,'end_diff'] = amy_prion_df_alt['end'].astype(int) + amy_prion_df_alt['seq_diff'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(187, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P01236</td>\n",
       "      <td>35</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P01236</td>\n",
       "      <td>71</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P01258</td>\n",
       "      <td>90</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P01258</td>\n",
       "      <td>99</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P02647</td>\n",
       "      <td>70</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id  begin  end\n",
       "0     P01236     35   62\n",
       "1     P01236     71   85\n",
       "2     P01258     90   95\n",
       "3     P01258     99  104\n",
       "4     P02647     70   83"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amy_prion_df_change = amy_prion_df_alt.drop(['begin','end','seq_diff'],axis=1)\n",
    "amy_prion_df_change.columns = ['uniprot_id','begin','end']\n",
    "print(amy_prion_df_change.shape)\n",
    "amy_prion_df_change.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect('human_protein.db')\n",
    "amy_prion_df_change.to_sql('amyloid_prion_id_mapped',con=conn,index=False,if_exists='replace')\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you're wondering how/why the amy_prion table expanded...in the amyloid database, if a protein is annotated as producing multiple amyloid proteins - they make a new entry for it in the database.\n",
    "\n",
    "My doing the above, I correctly mapped these extra amyloids to the original uniprot protein sequence. So my list of amyloid sections expanded due to those proteins having multiple entries."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse amyloids not yet present in protein db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect(\"human_protein.db\")\n",
    "with open('data/backfill_amyloids.xml') as f:\n",
    "    amyloid_protein_dict = xmltodict.parse(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "def commentAlgo(comment,locs,):\n",
    "    \"\"\"I'm having a problem with entries sometimes being lists or flat entries\"\"\"\n",
    "\n",
    "def entryCommentParser(entry_name, comments):\n",
    "    function_list = []\n",
    "    disease_list = []\n",
    "    tissue_list = []\n",
    "    subcellular_list = []\n",
    "    if type(comments)!=list:\n",
    "        comments = [comments]\n",
    "    for comment in comments:\n",
    "        comm_arr = [entry_name]\n",
    "        comm_type = comment.get(\"@type\",None)\n",
    "        value = comment.get(\"text\")\n",
    "        if comm_type in ['function','pathway','activity regulation','similarity']:\n",
    "            if value!=None and type(value)!=str:\n",
    "                value = value.get(\"#text\")\n",
    "                comm_arr.extend([comm_type, value])\n",
    "                function_list.append(comm_arr)\n",
    "        elif comm_type=='disease':\n",
    "            comm_disease = comment.get('disease')\n",
    "            if type(comm_disease)!='str' and comm_disease!=None:\n",
    "                comm_arr.append(comment.get('disease').get(\"name\"))\n",
    "                disease_list.append(comm_arr)\n",
    "        elif comm_type=='tissue specificity':\n",
    "            value = comment.get('text')\n",
    "            if type(value)!=str:\n",
    "                value = value.get(\"#text\")\n",
    "            comm_arr.append(value)\n",
    "            tissue_list.append(comm_arr)\n",
    "        # This is a really messily annotated section in these entries...sheesh!\n",
    "        elif comm_type=='subcellular location':\n",
    "            locs = []\n",
    "            loc_source = comment.get('subcellularLocation')\n",
    "            if type(loc_source)!=list:\n",
    "                loc_source = [loc_source]\n",
    "            for loc_head in loc_source:\n",
    "                if loc_head == None:\n",
    "                    pass\n",
    "                else:\n",
    "                    sub_locs = loc_head.get('location')\n",
    "                    if type(sub_locs)==list:\n",
    "                        for sub_sub_locs in sub_locs:\n",
    "                            if type(sub_sub_locs)==str:\n",
    "                                value = sub_sub_locs\n",
    "                            else:\n",
    "                                value = sub_sub_locs.get('#text')\n",
    "                            if value not in locs:\n",
    "                                locs.append(value)\n",
    "                                subcellular_list.append([entry_name, value])\n",
    "                    elif type(sub_locs)==str:\n",
    "                        value = sub_locs\n",
    "                    else:\n",
    "                        value = sub_locs.get('#text')\n",
    "                    if value not in locs:\n",
    "                        locs.append(value)\n",
    "                        subcellular_list.append([entry_name, value])\n",
    "        else:\n",
    "            pass\n",
    "    return function_list, disease_list, tissue_list, subcellular_list\n",
    "\n",
    "def featureParser(entry_name, features):\n",
    "    \"\"\"Right now I only want features that are annotations of;\n",
    "    \n",
    "    -strand (beta sheet)\n",
    "    -helix  (alpha helix)\n",
    "    -turn   (highly structured secondary structure)\n",
    "    \n",
    "    \"\"\"\n",
    "    sec_struc_list = []\n",
    "    modified_residues = []\n",
    "    feat_arr = []\n",
    "    for feature in features:\n",
    "        if type(feature)==str:\n",
    "            pass\n",
    "        else:\n",
    "            feat_type = feature.get(\"@type\")\n",
    "            if feat_type in ['strand','helix','turn']:\n",
    "                feat_locations = feature.get('location')\n",
    "                feat_begin = feat_locations.get('begin').get('@position')\n",
    "                feat_end = feat_locations.get('end').get('@position')\n",
    "                sec_struc_list.append([entry_name, feat_type, feat_begin, feat_end])\n",
    "            elif feat_type in ['modified residue','non-standard residue','lipidation',\n",
    "                               'glycosylation','disulfide bond','cross-link']:\n",
    "                feat_desc = feature.get(\"@description\")\n",
    "                feat_posi = feature.get(\"location\")\n",
    "                if 'begin' in feat_posi.keys():\n",
    "                    begin = feat_posi.get('begin').get(\"@position\")\n",
    "                    end = feat_posi.get('end').get(\"@position\")\n",
    "                    if begin==None:\n",
    "                        feat_posi = end\n",
    "                    elif end==None:\n",
    "                        feat_posi = begin\n",
    "                    elif begin==None and end==None:\n",
    "                        feat_posi = None\n",
    "                    else:\n",
    "                        feat_posi = begin+\"-\"+end\n",
    "                else:\n",
    "                    feat_posi = feat_posi.get('position').get(\"@position\")\n",
    "                modified_residues.append([entry_name, feat_type, feat_desc, feat_posi])\n",
    "\n",
    "    return sec_struc_list, modified_residues\n",
    "\n",
    "def keywordParser(keyword):\n",
    "    text = []\n",
    "    for x in keyword:\n",
    "        if type(x)==str:\n",
    "            text.append(x)\n",
    "        elif x!=None and type(x)!=str:\n",
    "            text.append(x.get('#text'))\n",
    "    return ', '.join(text)\n",
    "\n",
    "def entryParser(entry, conn, print_option=False):\n",
    "    # Get name\n",
    "    entry_name = entry.get('name')\n",
    "    # Get accession IDs\n",
    "    entry_accessions = entry.get('accession')\n",
    "    if type(entry_accessions)==list:\n",
    "        entry_accessions = \",\".join(entry_accessions)\n",
    "    # Get sequence\n",
    "    entry_seq = entry.get('sequence').get('#text').replace(\"\\n\",\"\").strip()\n",
    "    # Parse keywords\n",
    "    entry_keywords = \"\"\n",
    "    if entry.get('keyword')!=None:\n",
    "        entry_keywords = keywordParser(entry.get('keyword'))\n",
    "    # Parse features\n",
    "    entry_features = entry.get('feature')\n",
    "    if entry_features!=None:\n",
    "        sec_struc_list, modified_residues = featureParser(entry_name, entry_features)\n",
    "    else:\n",
    "        sec_struc_list = []\n",
    "        modified_residues = []\n",
    "    # Parse comments\n",
    "    entry_comments = entry.get('comment', None)\n",
    "    comm_list = []\n",
    "    disease_list = []\n",
    "    tissue_list = []\n",
    "    subcellular_list = []\n",
    "    if entry_comments != None:\n",
    "        comm_list, disease_list, tissue_list, subcellular_list \\\n",
    "        = entryCommentParser(entry_name, entry_comments)\n",
    "    \n",
    "    # One row each protein: protein name, sequence and keyword string\n",
    "    protein_df = pd.DataFrame([[entry_name, entry_accessions, entry_seq, entry_keywords]],\n",
    "                              columns=[\"protein\",\"accession\",\"sequence\",\"keywords\"])\n",
    "    # Multiple rows each protein for secondary structure: protein name, secondary structure type,\n",
    "    # where it starts, where it ends\n",
    "    sec_struc_df = pd.DataFrame(sec_struc_list,\n",
    "                                columns=[\"protein\",\"sec_struc_type\",\"begin\",\"end\"])\n",
    "    # Multiple rows each protein: protein name, modification type, description of modification,\n",
    "    # position of modification\n",
    "    aa_mod_df = pd.DataFrame(modified_residues,\n",
    "                             columns=['protein','modification','description','position'])\n",
    "    # Multiple rows each protein for comments: protein name, comment type, comment value\n",
    "    comment_df = pd.DataFrame(comm_list, columns=[\"protein\",\"comm_type\",\"value\"])\n",
    "    \n",
    "    # Multiple rows each protein for disease associations: protein name, disease name\n",
    "    disease_df = pd.DataFrame(disease_list, columns=[\"protein\",\"disease\"])\n",
    "    \n",
    "    # Multiple rows each protein for tissue expression: protein name, tissue\n",
    "    tissue_df = pd.DataFrame(tissue_list, columns=['protein','disease'])\n",
    "    \n",
    "    # Multiple rows each protein for subcellular localization: protein name, subcellular loc\n",
    "    subcellular_df = pd.DataFrame(subcellular_list, columns=['protein','subcellular_loc'])\n",
    "    \n",
    "    if print_option:\n",
    "        if protein_df.shape[0]!=0:\n",
    "            print(\"Protein DataFrame\")\n",
    "            display(protein_df)\n",
    "        if sec_struc_df.shape[0]!=0:\n",
    "            print(\"Secondary structure DataFrame\")\n",
    "            display(sec_struc_df)\n",
    "        if aa_mod_df.shape[0]!=0:\n",
    "            print(\"Amino Acid Modifications\")\n",
    "            display(aa_mod_df)\n",
    "        if comment_df.shape[0]!=0:\n",
    "            print(\"Comment DataFrame\")\n",
    "            display(comment_df)\n",
    "        if disease_df.shape[0]!=0:\n",
    "            print(\"Disease DataFrame\")\n",
    "            display(disease_df)\n",
    "        if tissue_df.shape[0]!=0:\n",
    "            print(\"Tissue DataFrame\")\n",
    "            display(tissue_df)\n",
    "        if subcellular_df.shape[0]!=0:\n",
    "            print(\"Subcellular Location DataFrame\")\n",
    "            display(subcellular_df)\n",
    "            \n",
    "    # Record entries in sqlite\n",
    "    if protein_df.shape[0]!=0:\n",
    "        protein_df.to_sql('protein',conn,if_exists='append',index=False)\n",
    "    if sec_struc_df.shape[0]!=0:\n",
    "        sec_struc_df.to_sql('protein_secondary_structure',conn,if_exists='append',index=False)\n",
    "    if aa_mod_df.shape[0]!=0:\n",
    "        aa_mod_df.to_sql('protein_amino_acid_modifications',conn,if_exists='append',index=False)\n",
    "    if comment_df.shape[0]!=0:\n",
    "        comment_df.to_sql('protein_comments',conn,if_exists='append',index=False)\n",
    "    if disease_df.shape[0]!=0:\n",
    "        disease_df.to_sql('protein_diseases',conn,if_exists='append',index=False)\n",
    "    if tissue_df.shape[0]!=0:\n",
    "        tissue_df.to_sql('protein_tissue_expression',conn,if_exists='append',index=False)\n",
    "    if subcellular_df.shape[0]!=0:\n",
    "        subcellular_df.to_sql('protein_subcellular_localization',conn,if_exists='append',index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0d9d3db6ea44486d9b537f30ab50fa8a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=70), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# \n",
    "for entry in tqdm_notebook(amyloid_protein_dict['uniprot']['entry']):\n",
    "    entryParser(entry,conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropped 63 rows from protein\n",
      "Dropped 499 rows from protein_secondary_structure\n",
      "Dropped 246 rows from protein_amino_acid_modifications\n",
      "Dropped 95 rows from protein_comments\n",
      "Dropped 4 rows from protein_diseases\n",
      "Dropped 9 rows from protein_tissue_expression\n",
      "Dropped 1084 rows from protein_subcellular_localization\n"
     ]
    }
   ],
   "source": [
    "# Drop duplicates in each table.\n",
    "for table in ['protein','protein_secondary_structure',\n",
    "              'protein_amino_acid_modifications', 'protein_comments',\n",
    "              'protein_diseases','protein_tissue_expression',\n",
    "              'protein_subcellular_localization']:\n",
    "    loaded_df = pd.read_sql(\"select * from \"+table,con=conn)\n",
    "    shape_bef = loaded_df.shape[0]\n",
    "    loaded_df.drop_duplicates(inplace=True)\n",
    "    shape_aft = loaded_df.shape[0]\n",
    "    if shape_bef > shape_aft:\n",
    "        print(\"Dropped \"+str(shape_bef-shape_aft)+\" rows from \"+table)\n",
    "    loaded_df.to_sql(table,con=conn,index=False,if_exists='replace')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def nGrammer(protein, seq, n):\n",
    "    return [[protein, i+1]+list(x) for i, x in enumerate(nltk.ngrams(seq, n))]\n",
    "\n",
    "def dfNGrammer(protein_df, num_grams):\n",
    "    gram_cols = [\"protein\",\"gram_num\"]+[\"gram_\"+str(i+1) for i in range(num_grams)]\n",
    "    seq_grams = []\n",
    "    seq_grams.extend([val for sublist in protein_df.apply(lambda row: nGrammer(row['protein'], row['sequence'], num_grams),axis=1) for val in sublist])\n",
    "    gram_df = pd.DataFrame(seq_grams, columns=gram_cols)\n",
    "    return gram_df\n",
    "        \n",
    "#ngram_df = dfNGrammer(prot_feat,5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prot_feat.shape,ngram_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ngram_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
